### PR TITLE
drivers: wifi: Use signal field from RX events

### PR DESCRIPTION
**Pull request by ajayparida**
For RSSI queries RPU only if no data is active.
If data is active maintains average signal strength and then provides it to WPA supplicant.

### CODE REVIEW

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_structs.h [hunk 1]
```diff
@@ -273,6 +273,8 @@ struct wifi_nrf_fmac_callbk_fns {
 					  unsigned short frequency,
 					  signed short signal);
 #endif /* CONFIG_WIFI_MGMT_RAW_SCAN_RESULTS */
+	void (*rx_rssi_info_fn)(void *os_vif_ctx,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_structs.h]: RSSI implies RX.
```suggestion
	void (*process_rssi)(void *os_vif_ctx,
```

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h [hunk 2]
```diff
@@ -36,6 +36,8 @@
  * connected scan.
  */
 #define WIFI_NRF_SCAN_TIMEOUT (K_SECONDS(30))
+#define MBM_TO_DBM 100
+#define MAX_RSSI_RECORD_TIME_GAP_MS 100 /* ms */
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h]:
```suggestion
#define RSSI_STALE_TIMEOUT_MS 100
```

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c [hunk 3]
```diff
@@ -1004,32 +1004,25 @@ int wifi_nrf_wpa_supp_signal_poll(void *if_priv, struct wpa_signal_info *si, uns
...
-		goto out;
-	}
-
-	ret = wifi_nrf_fmac_get_interface(rpu_ctx_zep->rpu_ctx, vif_ctx_zep->vif_idx);
-	if (ret != WIFI_NRF_STATUS_SUCCESS) {
-		LOG_ERR("%s: Failed to get interface info\n", __func__);
-		goto out;
-	}
+	if ((wifi_nrf_osal_time_elapsed_us(fmac_dev_ctx->fpriv->opriv,
+		vif_ctx_zep->rssi_record_timestamp_us) / 1000) > MAX_RSSI_RECORD_TIME_GAP_MS) {
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c]: Shouldn't this be reverse? If the time elapsed is > MAX_RSSI_RECORD_TIME_GAP_MS then its time for us to query RPU

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c [hunk 4]
```diff
@@ -1004,32 +1004,25 @@ int wifi_nrf_wpa_supp_signal_poll(void *if_priv, struct wpa_signal_info *si, uns
...
-		ret = WIFI_NRF_STATUS_FAIL;
-		goto out;
+		sem_ret = k_sem_take(&wait_for_event_sem, K_MSEC(RPU_RESP_EVENT_TIMEOUT));
+		if (sem_ret) {
+			LOG_ERR("%s: Failed to get station info, ret = %d\n", __func__, sem_ret);
+			ret = WIFI_NRF_STATUS_FAIL;
+			goto out;
+		}
 	}
-	vif_ctx_zep->signal_info->frequency = vif_ctx_zep->assoc_freq;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c]: Why is this removed?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c [hunk 5]
```diff
@@ -1004,32 +1004,25 @@ int wifi_nrf_wpa_supp_signal_poll(void *if_priv, struct wpa_signal_info *si, uns
...
-		goto out;
-	}
+	if ((wifi_nrf_osal_time_elapsed_us(fmac_dev_ctx->fpriv->opriv,
+		vif_ctx_zep->rssi_record_timestamp_us) / 1000) > MAX_RSSI_RECORD_TIME_GAP_MS) {
+		si->current_signal = (int)vif_ctx_zep->rssi;
+		ret = WIFI_NRF_STATUS_SUCCESS;
+	} else {
+		ret = wifi_nrf_fmac_get_station(rpu_ctx_zep->rpu_ctx, vif_ctx_zep->vif_idx, bssid);
+		if (ret != WIFI_NRF_STATUS_SUCCESS) {
+			LOG_ERR("%s: Failed to get station info\n", __func__);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c]:
```suggestion
			LOG_ERR("%s: Failed to send get station info command\n", __func__);
```

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c (hunk 3 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c]: Yes it should be

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c (hunk 4 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c]: It is required, Mistakenly removed.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c (hunk 5 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/zephyr/src/zephyr_wpa_supp_if.c]: I have updated. Please check.

**Comment by ajayparida** [comment]
Local PR not needed any more as it's handled in https://github.com/nrfconnect/sdk-nrf/pull/11510
